### DISCUSSION TITLE
Provide capability to use multiple configuration files for monorepos

### DISCUSSION THREAD

**Discussion by tomkerkhove** [renovatebot/renovate#19721] [Suggest an Idea]
### What would you like Renovate to be able to do?

Provide capability to use multiple configuration files for monorepos, so that per component we can have a Renovate configuration to keep them small and scoped.

While re-usable presets can help; top-level configuration such as `includePaths` is not supported.

Originates from https://github.com/renovatebot/renovate/discussions/19666.

### If you have any ideas on how this should be implemented, please tell us here.

N/A

### Is this a feature you are interested in implementing yourself?

No

**Comment by viceice** [comment]
unlikely to happen

**Comment by rarkins** [comment]
I'm not sure how this would even work. More details needed

**Reply by tomkerkhove** [reply]
So here is what we have now and is not really working:

```json
{
  "$schema": "https://docs.renovatebot.com/renovate-schema.json",
  "extends": [
    "One/repo//Automation/Renovate/config/renovate.component-a",
    "One/repo//Automation/Renovate/config/renovate.component-b"
  ]
}
```

These are respective configurations for a single component in a mono-repo and is tailored to that repo. Some examples are top-level configuration options such as `branchPrefix`, `labels`, `commitMessageSuffix`, `enabledManagers` & `includePaths`.

When running the above, I'm seeing that Renovate will use the top-level configuration from component-b and ignore component-a, which is probably by design? Is there a way that it just runs both in parallel?

In theory, I could move `includePaths` et al to package rules but that introduces a lot of duplication. 

I was hoping to copy over the appropriate configuration to `renovate.json` but it looks like Renovate is doing a full clone of the repo so that is not working either. Or is there a way to avoid cloning the repo and run in current folder? I see the folders are [documented](https://docs.renovatebot.com/examples/self-hosting/#filedirectory-usage) but likely it will be overridden anyway

**Comment by tomkerkhove** [comment]
I'm not a TS/JS expert, but it is currently ran by using `npx --userconfig .npmrc renovate` and was wondering if we could add a configuration similar to `--config <path>`.

**Comment by rarkins** [comment]
Renovate operates on multiple repos per run. If you want a different global config like your example then you can already do that

**Comment by tomkerkhove** [comment]
Just to be clear - I'm asking for a self-hosted scenario.

But that's what I figured; I'm curious to know how it is running them? Imagine I have 1 repo with 2 configs; how do I tell it to use config A and not B?

**Reply by tomkerkhove** [reply]
> Renovate operates on multiple repos per run. If you want a different global config like your example then you can already do that

This is OK; but is the above scenario also supported @rarkins?

**Reply by rarkins** [reply]
I don't understand your question

**Reply by viceice** [reply]
you can only have one config per repo.

**Reply by tomkerkhove** [reply]
And that's the feature request I had before it turned in to a discussion :)

But since it's a self-hosted solution we went another route and rename configurations before running Renovate.

So `renovate.{}.json` -> `renovate.json`, but it would be useful to have this supported out-of-the-box for monorepos so that components can have difference configurations.

**Reply by viceice** [reply]
nope, you should use package rules for specific settings. you can spread them into multiple presets to have multiple smaller files.

**Reply by viceice** [reply]
you also still didn't explained in detail what you try to solve with multiple config files. i don't see any options which can't be used in a single file besides some which already have open issues.

**Reply by tomkerkhove** [reply]
The problem is that not everything is defined with package rules or you don't want to duplicate it for every single rule in a file; so you cannot fully rely on that.

Examples of what we use on top-level (either to remove duplication or because only place available):
- `includePaths`
- `enabledManagers`

Note that we also use a "disable by default, enable what we want" approach.

For example:
```json
"packageRules": [
    {
      "groupName": "Ignore NuGet updates out-of-the-box",
      "matchDatasources": ["nuget"],
      "excludePackagePrefixes": [
        // Packages we care about
      ],
      "enabled": false
    },
    {
      "description": "Disable major NuGet updates",
      "matchDatasources": ["nuget"],
      "matchUpdateTypes": ["major"],
      "enabled": false
    },
    {
      "groupName": "Docker images",
      "matchDatasources": ["docker"]
    },
    {
      "groupName": "Other (.NET)",
      "matchDatasources": ["nuget"],
      "matchUpdateTypes": ["minor", "patch"],
      "labels": [ "Dependencies" ]
    },
    // ...
]
```

**Reply by viceice** [reply]
you can extend presets inside package rules.
so create presets for dedupe.

running multiple configs  on a single repo doesn't work, as renovate would drop previous branches with update which doesn't belong to the current config.

**Reply by tomkerkhove** [reply]
Do you have an example of what you mean please? I'm not looking to reduce duplication of the package rules themselves.

**Reply by tomkerkhove** [reply]
Ok my approach did not work as Renovate clones the whole repo - Is there a way not to clone the repo as part of Renovate and instead just rely on the local files?

When running this in pipelines (such as Azure Pipelines) the content is already there and would be nice to just run it based on that.

**Reply by viceice** [reply]
no, not possible.

**Reply by tomkerkhove** [reply]
That's fine, but has this been taken in to consideration?

**Reply by viceice** [reply]
no, doesn't fit to renovate workflow

**Reply by tomkerkhove** [reply]
It does for self-hosted scenario.

**Reply by viceice** [reply]
i mean it doesn't fit to current internal renovate architecture.

**Comment by rarkins** [comment]
Converted to a discussion

**Comment by gberche-orange** [comment]
My team is meeting similar challenges with a single renovate.json5 file configured into our mono repo:
* the renovate.json5 file is now 350 lines long and hard to navigate 
* git conflicts occur with multiple contributions to the same file
* with many uses of regexp manager, it is common for a given dependency to require changes to the `packageRules` and `regexManagers`,  which become long, without a clear sorting order

As I understand github hosted presets from https://docs.renovatebot.com/config-presets/#github-hosted-presets , extracting parts of configuration from our github>org/repo:renovate.json5 into github>org/repo:component1.json5 and github>org/repo:component2.json5 has the following drawbacks for the usage of mend bot on github:
- the `"extends": ["github>org/repo:component1.json5"]` needs to be modified in each clone, or [minimal reproduction repos](https://github.com/renovatebot/renovate/blob/main/docs/development/minimal-reproductions.md) to point to the local repo
- requires explicitly listing every component

The modularity issue in configuration seems a common problem to many systems. This might offer inspiration and reuse for renovate ?

For instance, K8S and kustomize use [strategic-merge-patch](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-api-machinery/strategic-merge-patch.md) or Json patches, see https://kubectl.docs.kubernetes.io/references/kustomize/kustomization/patches/

This enables granular fine-grain, focussed files contributing to constructing a single large (json) document.

Could renovate support similar patches applying on the current single configuration file ?

**Reply by rarkins** [reply]
Are you looking to have multiple configuration files which you switch between, or rather the ability for a renovate.json to include e.g. `backend/renovate.json` and `frontend/renovate.json` so that teams can manage their own config (and hopefully don't contradict each other..)

**Reply by tomkerkhove** [reply]
Our main struggle is that not everything is supported when using `backend/renovate.json` & `frontend/renovate.json` approach. In terms of switching is also not an alternative for us, we tried to use that route but renovate always clones the repo and instantly runs so you cannot rename files or so.

**Reply by gberche-orange** [reply]
I'm rather looking for the ability for a renovate.json to include e.g. backend/renovate.json and frontend/renovate.json. In my particular use-case, this would rather be worded as type of artefacts such as bosh-releases.json5 (for [bosh releases](https://bosh.io/docs/release/)) bosh-stemcells.json5, crossplane.json5, vendir.json5 etc...

Ideally, the top renovate.json5 could include a regexp of files to load from so that the addition of a new file does not require modification of the root renovate5.json

Regarding conflicts, documented precedence rules would indicate what to expect when loading the additional files. Also the renovate dashboard would have debug traces displaying the resulting final configuration.

**Reply by Memfisrain** [reply]
@rarkins Hey, we're in the process of researching a solution for automatic dependency update for our company and renovate looks pretty solid and configurable solution.

Could you update on the status of the question above, like the situation when we have a monorepo with _frontend_ and _backend_ packages and would like to include _backend/renovate.json_ and _frontend/renovate.json_.

Anyway, thanks for the great product.

**Reply by rarkins** [reply]
A simple includes could be achieved today with "local>" presets pointing to the same repo. Anything advanced isn't planned unless a Mend.io enterprise customer needs it.

**Reply by Memfisrain** [reply]
got it, thanks!
